Zotero/Scenes/Detail/Lookup/Views/LookupViewController.swift

```diff
@@ -100,124 +102,139 @@ class LookupViewController: UIViewController {
-                guard let self = self else { return }
-                self.activityIndicator.stopAnimating()
-                self.activityIndicator.isHidden = true
-                self.errorLabel.isHidden = true
-                self.tableView.isHidden = false
-
-                self.dataReloaded?()
-
-                self.closeAfterUpdateIfNeeded()
+            show(data: data) { [weak self] in
+                guard let self else { return }
+                activityIndicator.stopAnimating()
+                activityIndicator.isHidden = true
+                errorLabel.isHidden = true
+                tableView.isHidden = false
+                dataReloaded?()
+                closeAfterUpdateIfNeeded()
             }
         }
     }
 
     private func show(data: [LookupState.LookupData], completion: @escaping () -> Void) {
-        var snapshot = NSDiffableDataSourceSnapshot<Int, Row>()
-        snapshot.appendSections([0])
-
-        for lookup in data {
-            switch lookup.state {
-            case .translated(let translationData):
-                let title: String
-                if let _title = translationData.response.fields[KeyBaseKeyPair(key: FieldKeys.Item.title, baseKey: nil)] {
-                    title = _title
-                } else {
-                    let _title = translationData.response.fields.first(where: { self.schemaController.baseKey(for: translationData.response.rawType, field: $0.key.key) == FieldKeys.Item.title })?.value
-                    title = _title ?? ""
-                }
-                let itemData = Row.Item(type: translationData.response.rawType, title: title)
-
-                snapshot.appendItems([.item(itemData)], toSection: 0)
+        tableView.isHidden = false
 
-                let attachments = translationData.attachments.map({ attachment -> Row in
-                    let (progress, error) = self.remoteFileDownloader.data(for: attachment.0.key, parentKey: translationData.response.key, libraryId: attachment.0.libraryId)
-                    let updateKind: RemoteAttachmentDownloader.Update.Kind
-                    if error != nil {
-                        updateKind = .failed
-                    } else if let progress = progress {
-                        updateKind = .progress(progress)
-                    } else {
-                        updateKind = .ready(attachment.0)
-                    }
-                    return .attachment(attachment.0, updateKind)
-                })
-                snapshot.appendItems(attachments, toSection: 0)
+        updateQueue.async { [weak self] in
+            guard let self else { return }
 
-            case .failed:
-                snapshot.appendItems([.identifier(identifier: lookup.identifier, state: .failed)])
+            let snapshot = createSnapshot(from: data, schemaController: schemaController, remoteFileDownloader: remoteFileDownloader)
+            dataSource.apply(snapshot, animatingDifferences: false)
```

Comment:
The following completion can be written using the completion closure of `apply`, which is called in the main queue by default as such (I write as a suggestion for just this line, because the overlapping deleted lines don't allow a suggestion):
```suggestion
            dataSource.apply(snapshot, animatingDifferences: false) { [weak self] in
                guard let self, contentSizeObserver == nil else {
                    completion()
                    return
                }

                // For some reason, the observer subscription has to be here, doesn't work if it's in `viewDidLoad`.
                contentSizeObserver = tableView.observe(\.contentSize, options: [.new]) { [weak self] _, change in
                    guard let self, let value = change.newValue, value.height != tableViewHeight.constant else { return }
                    tableViewHeight.constant = value.height
                    if value.height >= self.tableView.frame.height, !tableView.isScrollEnabled {
                        tableView.isScrollEnabled = true
                    }
                    completion()
                }
            }
```

Reply:
Good point, fixed in e3401b2c8d91ce78f68b9009c829881eff063683